File: backend/speed_reading/progression.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class ProgressionLevel:
    id: int
    world: int
    step: int
    words_per_chunk: int
    wpm: int
    pass_threshold: int


@dataclass(frozen=True)
class World:
    world: int
    words_per_chunk: int
    name: str


def _root() -> Path:
    return Path(__file__).resolve().parents[2]


def load_progression(path: Path | None = None) -> tuple[list[World], list[ProgressionLevel]]:
    source = path or _root() / "data" / "progression.json"
    raw = json.loads(source.read_text(encoding="utf-8"))
    worlds = [
        World(world=item["world"], words_per_chunk=item["wordsPerChunk"], name=item["name"])
        for item in raw["worlds"]
    ]
    # SR-R1-011: the level ladder is derived from the speedSteps/passThreshold configuration,
    # not read from a separately hard-coded/duplicated levels list - change the config, and the
    # eligible speeds change with it, no code change required.
    levels = build_levels_from_ladder(worlds, raw["speedSteps"], raw["passThreshold"])
    return worlds, levels


def build_levels_from_ladder(
    worlds: list[World], speed_steps: list[int], pass_threshold: int
) -> list[ProgressionLevel]:
    levels: list[ProgressionLevel] = []
    for world in worlds:
        for step_index, wpm in enumerate(speed_steps, start=1):
            levels.append(
                ProgressionLevel(
                    id=len(levels) + 1,
                    world=world.world,
                    step=step_index,
                    words_per_chunk=world.words_per_chunk,
                    wpm=wpm,
                    pass_threshold=pass_threshold,
                )
            )
    return levels


def build_expected_levels(
    speed_steps: list[int] | None = None,
    max_words_per_chunk: int = 6,
    pass_threshold: int = 70,
) -> list[ProgressionLevel]:
    """Canonical progression: every chunk size climbs the full speed ladder.

    World 1 highlights one word at a time from 100 to 200 WPM; each later world
    repeats the same ladder with one more word per chunk, up to six words.
    """
    steps = speed_steps or [100, 120, 140, 160, 180, 200]
    levels: list[ProgressionLevel] = []
    for chunk in range(1, max_words_per_chunk + 1):
        for index, wpm in enumerate(steps, start=1):
            levels.append(
                ProgressionLevel(
                    id=len(levels) + 1,
                    world=chunk,
                    step=index,
                    words_per_chunk=chunk,
                    wpm=wpm,
                    pass_threshold=pass_threshold,
                )
            )
    return levels


def stars_for_score(score: int) -> int:
    if score >= 90:
        return 3
    if score >= 80:
        return 2
    if score >= 70:
        return 1
    return 0


def unlocked_level_ids(levels: list[ProgressionLevel], passed_level_ids: set[int]) -> set[int]:
    """A level is unlocked when it is the first level or its predecessor is passed."""
    unlocked: set[int] = set()
    for level in levels:
        if level.id == 1 or (level.id - 1) in passed_level_ids:
            unlocked.add(level.id)
    return unlocked


def next_level_id(levels: list[ProgressionLevel], passed_level_ids: set[int]) -> int | None:
    """The next playable level: first unlocked level that is not yet passed."""
    unlocked = unlocked_level_ids(levels, passed_level_ids)
    for level in levels:
        if level.id in unlocked and level.id not in passed_level_ids:
            return level.id
    return None

File: lib/progression.ts
import progressionData from "../data/progression.json";
import { ReadingTimingRecord } from "./reading-timing";

export type ProgressionLevel = {
  id: number;
  world: number;
  step: number;
  wordsPerChunk: number;
  wpm: number;
  passThreshold: number;
};

export type WorldInfo = {
  world: number;
  wordsPerChunk: number;
  name: string;
};

export type LevelResult = {
  bestScore: number;
  stars: number;
  passed: boolean;
  completedAt: string;
  // SR-R1-001: raw evidence for the most recent attempt's reading phase - planned timing is
  // deterministic from target_wpm/word_count/chunks; actual_duration_ms is the real elapsed time.
  lastReadingTiming?: ReadingTimingRecord;
};

export type Progress = Record<string, LevelResult>;

export const worlds: WorldInfo[] = progressionData.worlds;

// SR-R1-011: Configurable WPM ladder.
// "Use a file/config-based WPM ladder rather than hard-coded values." The level ladder is
// derived from worlds/speedSteps/passThreshold at load time - never a separately hard-coded or
// duplicated list - so changing the config changes eligible speeds without a code change.
export function buildLevelsFromLadder(
  ladderWorlds: WorldInfo[],
  speedSteps: number[],
  passThreshold: number
): ProgressionLevel[] {
  const built: ProgressionLevel[] = [];
  for (const world of ladderWorlds) {
    speedSteps.forEach((wpm, index) => {
      built.push({
        id: built.length + 1,
        world: world.world,
        step: index + 1,
        wordsPerChunk: world.wordsPerChunk,
        wpm,
        passThreshold
      });
    });
  }
  return built;
}

export const levels: ProgressionLevel[] = buildLevelsFromLadder(
  progressionData.worlds,
  progressionData.speedSteps,
  progressionData.passThreshold
);

const STORAGE_KEY = "speedreader-progress-v1";

// When Speed Reading is opened from inside BabySteps, progress is namespaced by learnerId so
// siblings sharing a device (or the browser's own anonymous progress) never collide.
function storageKey(learnerId?: string): string {
  return learnerId ? `${STORAGE_KEY}:${learnerId}` : STORAGE_KEY;
}

export function loadProgress(learnerId?: string): Progress {
  if (typeof window === "undefined") return {};
  try {
    const raw = window.localStorage.getItem(storageKey(learnerId));
    return raw ? (JSON.parse(raw) as Progress) : {};
  } catch {
    return {};
  }
}

export function saveProgress(progress: Progress, learnerId?: string): void {
  if (typeof window === "undefined") return;
  window.localStorage.setItem(storageKey(learnerId), JSON.stringify(progress));
}

export function starsForScore(score: number): number {
  if (score >= 90) return 3;
  if (score >= 80) return 2;
  if (score >= 70) return 1;
  return 0;
}

export function recordResult(
  progress: Progress,
  level: ProgressionLevel,
  score: number,
  learnerId?: string,
  readingTiming?: ReadingTimingRecord,
  attemptOutcome?: "PASS" | "HOLD" | "INVALID"
): Progress {
  // SR-R1-012: Outcome separation - technical/content invalidity must never lower (or otherwise
  // change) learner state the way a genuine comprehension failure (HOLD) does.
  if (attemptOutcome === "INVALID") {
    return progress;
  }

  const key = String(level.id);
  const previous = progress[key];
  const bestScore = Math.max(previous?.bestScore ?? 0, score);
  const next: Progress = {
    ...progress,
    [key]: {
      bestScore,
      stars: starsForScore(bestScore),
      passed: bestScore >= level.passThreshold,
      completedAt: new Date().toISOString(),
      lastReadingTiming: readingTiming ?? previous?.lastReadingTiming
    }
  };
  saveProgress(next, learnerId);
  return next;
}

export function isPassed(progress: Progress, levelId: number): boolean {
  return progress[String(levelId)]?.passed ?? false;
}

export function isUnlocked(progress: Progress, level: ProgressionLevel): boolean {
  return level.id === 1 || isPassed(progress, level.id - 1);
}

export function nextPlayableLevel(progress: Progress): ProgressionLevel | null {
  for (const level of levels) {
    if (isUnlocked(progress, level) && !isPassed(progress, level.id)) {
      return level;
    }
  }
  return null;
}

// SR-R1-011: current_rate/next_rate - the WPM the learner is eligible to train at right now,
// and the rate one step ahead on the ladder, derived from the same configured ladder.
export function currentRate(progress: Progress): number | null {
  return nextPlayableLevel(progress)?.wpm ?? null;
}

export function nextRate(progress: Progress): number | null {
  const current = nextPlayableLevel(progress);
  if (!current) return null;
  return levels.find((level) => level.id === current.id + 1)?.wpm ?? null;
}

export function passedCount(progress: Progress): number {
  return levels.filter((level) => isPassed(progress, level.id)).length;
}

export function totalStars(progress: Progress): number {
  return levels.reduce((sum, level) => sum + (progress[String(level.id)]?.stars ?? 0), 0);
}
